sdk/rust/oak/src/grpc/mod.rs

```diff
@@ -295,74 +295,71 @@ where
-        Ok(rsp) => writer.write(&rsp, WriteMode::Close),
-        Err(status) => writer.close(Err(status)),
-    };
-    if let Err(e) = result {
-        panic!("Failed to process response: {:?}", e)
-    }
+/// Trait implemented by all server method variants, applied to a server receiver instance having
+/// concrete type `T`.
+pub trait Invocable<T> {
+    /// Generic function that handles request deserialization and response serialization (and
+    /// sending down a channel) for a function that handles a request / response pair (either unary
+    /// or streaming).
+    ///
+    /// Panics if [de-]serialization or channel operations fail.
+    fn invoke(&self, target: &mut T, req: &[u8], writer: ChannelResponseWriter);
 }
 
-/// Generic function that handles request deserialization and response
-/// serialization (and sending down a channel) for a function that handles a
-/// request and streams responses.
+/// Enum with variants for all combinations of {request, response} × {unary , streaming}, each
+/// containing a pointer to the function of the appropriate type defined on a service
+/// implementation.
 ///
-/// Panics if [de-]serialization or channel operations fail.
-pub fn handle_req_stream<C, R>(mut node_fn: C, req: &[u8], writer: ChannelResponseWriter)
-where
-    C: FnMut(R, ChannelResponseWriter),
-    R: prost::Message + Default,
-{
-    let r = R::decode(req).expect("Failed to parse request protobuf message");
-    node_fn(r, writer)
-}
-
-/// Generic function that handles request deserialization and response
-/// serialization (and sending down a channel) for a function that handles a
-/// stream of requests to produce a response.
+/// The functions are "uncurried", i.e. they include an explicit parameter for the receiver type,
+/// which corresponds to `self` in the method definition.
 ///
-/// Panics if [de-]serialization or channel operations fail.
-pub fn handle_stream_rsp<C, R, Q>(mut node_fn: C, req: &[u8], writer: ChannelResponseWriter)
-where
-    C: FnMut(Vec<R>) -> Result<Q>,
-    R: prost::Message + Default,
-    Q: prost::Message + Default,
-{
-    // TODO(#97): better client-side streaming
-    let rr = vec![R::decode(req).expect("Failed to parse request protobuf message")];
-    let result = match node_fn(rr) {
-        Ok(rsp) => writer.write(&rsp, WriteMode::Close),
-        Err(status) => writer.close(Err(status)),
-    };
-    if let Err(e) = result {
-        panic!("Failed to process response: {:?}", e)
-    }
+/// Note that the variants with response stream do not make use of the type parameter `Q`, so this
+/// must be manually specified when creating one of those instances if the compiler cannot infer it
+/// from the context.
+pub enum ServerMethod<T, R, Q> {
+    UnaryUnary(fn(&mut T, R) -> Result<Q>),
+    UnaryStream(fn(&mut T, R, ChannelResponseWriter)),
+    StreamUnary(fn(&mut T, Vec<R>) -> Result<Q>),
+    StreamStream(fn(&mut T, Vec<R>, ChannelResponseWriter)),
 }
```

Comment:
I wonder if the `enum` is completely necessary &ndash; would it be possible to provide different `impl`s of the `Invocable<T>` trait for 4 different types:
 - `FnMut(&mut T, R) -> Result<Q>`
 - `FnMut(&mut T, R, ChannelResponseWriter)`
 - `FnMut(&mut T, Vec<R>) -> Result<Q>`
 - `FnMut(&mut T, Vec<R>, ChannelResponseWriter`

so that the right implementation gets picked via the trait object vtable rather than explicit `enum`?

(Again, could be a different PR)

Reply:
I had something like that in the process of writing this PR, but it ends up needing extra heap allocations, since the return value of the `server_method` would have to be something like `Option<Box<dyn Invocable<T>>>`.

The current version instead is simpler, it returns a (effectively static) reference to a dynamically dispatched struct: `Option<&dyn #oak_package::grpc::Invocable<T>>` , which makes sense, since the list of variants is known at compile time, so the boxing is unnecessary.

Or were you thinking of something else? Let me know if you have any other ideas.

Reply:
No, you're right that what I was thinking of would need a `Box::new` along the way.